File: qc-application/src/main/java/com/eager/questioncloud/application/business/user/service/UserAccountService.java
package com.eager.questioncloud.application.business.user.service;

import com.eager.questioncloud.application.mail.EmailSender;
import com.eager.questioncloud.core.domain.user.infrastructure.repository.UserRepository;
import com.eager.questioncloud.core.domain.user.model.User;
import com.eager.questioncloud.core.domain.verification.enums.EmailVerificationType;
import com.eager.questioncloud.core.domain.verification.implement.EmailVerificationProcessor;
import com.eager.questioncloud.core.domain.verification.model.Email;
import com.eager.questioncloud.core.domain.verification.model.EmailVerification;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserAccountService {
    private final UserRepository userRepository;
    private final EmailVerificationProcessor emailVerificationProcessor;
    private final EmailSender emailSender;

    public String recoverForgottenEmail(String phone) {
        User user = userRepository.getUserByPhone(phone);
        return user.getUserInformation().getEmail();
    }

    public void sendRecoverForgottenPasswordMail(String email) {
        User user = userRepository.getUserByEmail(email);
        EmailVerification emailVerification = emailVerificationProcessor.createEmailVerification(
            user.getUid(),
            email,
            EmailVerificationType.ChangePassword);
        emailSender.sendMail(Email.of(emailVerification));
    }

    public void sendChangePasswordMail(User user) {
        EmailVerification emailVerification = emailVerificationProcessor.createEmailVerification(
            user.getUid(),
            user.getUserInformation().getEmail(),
            EmailVerificationType.ChangePassword);
        emailSender.sendMail(Email.of(emailVerification));
    }

    @Transactional
    public void changePassword(String token, String newPassword) {
        EmailVerification emailVerification = emailVerificationProcessor.verifyEmailVerification(token, EmailVerificationType.ChangePassword);
        User user = userRepository.getUser(emailVerification.getUid());
        user.changePassword(newPassword);
        userRepository.save(user);
    }
}

File: qc-api/qc-user-api/src/main/kotlin/com/eager/questioncloud/user/account/service/UserAccountService.kt
package com.eager.questioncloud.user.account.service

import com.eager.questioncloud.user.account.implement.UserAccountUpdater
import com.eager.questioncloud.user.infrastructure.repository.UserRepository
import com.eager.questioncloud.user.mail.EmailSender
import com.eager.questioncloud.verification.domain.Email
import com.eager.questioncloud.verification.enums.EmailVerificationType
import com.eager.questioncloud.verification.implement.EmailVerificationProcessor
import org.springframework.stereotype.Service

@Service
class UserAccountService(
    private val userRepository: UserRepository,
    private val emailVerificationProcessor: EmailVerificationProcessor,
    private val userAccountUpdater: UserAccountUpdater,
    private val emailSender: EmailSender,
) {
    fun recoverForgottenEmail(phone: String): String {
        val user = userRepository.getUserByPhone(phone)
        return user.userInformation.email
    }

    fun sendRecoverForgottenPasswordMail(email: String) {
        val user = userRepository.getUserByEmail(email)
        val emailVerification = emailVerificationProcessor.createEmailVerification(
            user.uid,
            email,
            EmailVerificationType.ChangePassword
        )
        emailSender.sendMail(Email.of(emailVerification))
    }

    fun sendChangePasswordMail(userId: Long) {
        val user = userRepository.getUser(userId)
        val emailVerification = emailVerificationProcessor.createEmailVerification(
            user.uid,
            user.userInformation.email,
            EmailVerificationType.ChangePassword
        )
        emailSender.sendMail(Email.of(emailVerification))
    }

    fun changePassword(token: String, newPassword: String) {
        val emailVerification =
            emailVerificationProcessor.verifyEmailVerification(token, EmailVerificationType.ChangePassword)
        userAccountUpdater.changePassword(emailVerification.uid, newPassword)
    }
}
